### DISCUSSION TITLE
Web support

### DISCUSSION THREAD

**Discussion by mziadd** [simolus3/drift#1756] [Q&A]
Hi and thanks for this awesome library, We are using this library in our app on iOS and Android, and now we want to support the WEB platform, is it recommended to use this library for production use in case we support WEB?

**Comment by simolus3** [comment]
Drift supports running on the web through the sql.js library, this is described in the [documentation](https://drift.simonbinder.eu/web/). This approach has some challenges related to persistence, but it works for most users and is stable. There are apps using this approach in production. So when porting an app using drift to the web, this is the approach I recommend.

I'm also working on a new and more efficient web backend for drift that doesn't require an external JavaScript library. But this will take time to get stable, so using the approach in the documentation is the way to go for now.

**Comment by Jordan-Nelson** [comment]
Hello @simolus3. I had a few questions about web support now that some time has passed since this original question/answer.
- The docs note that web support is considered experimental (even when using sql.js). I see you noted that several apps are using this in production though. Is there anything in particular keeping this approach from being considered fully stable (known issues, limitations, planned API changes or deprecations)?
- Is the new approach that does not use external JS still considered experimental? Same question as above - Is there anything in particular keeping this approach from being considered stable (issues, limitations, etc.)?
- Is the long term plan to drop support for sql.js in favor of the approach that does not use JS?

Since the project I am working on is still in development, I likely will not need a production ready solution for a few months and would be happy to experiment with the new approach for the time being, but I wanted to get an idea of the current state of the two options and the direction moving forward.

Thank you for providing this package for the community.

**Reply by simolus3** [reply]
I don't have any breaking changes planned for either of the two web implementations. The main reason for them being considered experimental is that the persistence approach is still a bit hacky for both of them.

- sql.js has no builtin way to persist a database. The solution they recommend, which is also what drift is doing, is to export the entire database as a blob into memory and then persist that. We need to do this after every write. As you can imagine, this is not too performant and it doesn't scale to larger databases.
- In the new approach, we save the database to IndexedDb only (the old one also supports local storage or custom storage implementations), storing the file in 4kb chunks that can be updated independently. This, along with some caching, makes the implementation a lot more efficient. However, it is also a lot more complicated and there may still be some synchronization bugs which can lead to a corrupt database. There's also a fundamental limitation because, being a C library, sqlite3 expects synchronous access to a file system and we can only emulate async writes to IndexedDb. So, you loose the durability of ACID.

Long story short: I don't want to declare them production-ready as long as they have known persistence issues, and it's really hard to properly fix them. In fact, I've not seen a fully stable sqlite3-on-the-web solution without these problems yet (absurd sql comes a bit closer than our solution, but IIRC you can also break their persistence approach if the tab is closed in just the right moment). I know that the sqlite3 team is working on an official wasm port, so it will be interesting to see what they manage to come up with.

> Is the long term plan to drop support for sql.js in favor of the approach that does not use JS?

I'm pretty sure that one of them will be deprecated in favor of the other at some point, and I think the approach without external JS has more potential. Given the pain of migrating between them, my current plan is to keep both of them maintained though.

------

If you don't need local writes (for instance because you're always fetching a database from the server, or if you're only using the database as an in-memory cache), I don't see any problem with the new wasm approach.
